proof/refine/ARM/KHeap_R.thy

```diff
@@ -3921,4 +3927,259 @@ lemma threadGet_getObject:
+  apply (clarsimp simp: obj_at'_def state_refs_of'_def projectKOs)
+  done
+
+lemma SCReply_state_refs_of':
+  "(r, SCReply) \<in> state_refs_of' s p = obj_at' (\<lambda>ko. scReply ko = Some r) p s"
+  apply (rule iffI)
+   apply (clarsimp simp: state_refs_of'_def split: option.splits if_splits)
+   apply (rename_tac ko, case_tac ko; simp add: get_refs_def2)
+     apply (fastforce dest: prod_in_refsD)+
+   apply (clarsimp simp: obj_at'_def projectKOs tcb_bound_refs'_def get_refs_def2)
+  apply (clarsimp simp: obj_at'_def state_refs_of'_def projectKOs)
+  done
+
+lemma ReplySchedContext_state_refs_of':
+  "(scp, ReplySchedContext) \<in> state_refs_of' s p = obj_at' (\<lambda>ko. replyNext ko = Some (Head scp)) p s"
+  apply (rule iffI)
+   apply (clarsimp simp: state_refs_of'_def split: option.splits if_splits)
+   apply (rename_tac ko, case_tac ko; simp add: get_refs_def2)
+     apply (fastforce dest: prod_in_refsD)+
+   apply (clarsimp simp: obj_at'_def projectKOs tcb_bound_refs'_def get_refs_def2)
+  apply (clarsimp simp: obj_at'_def state_refs_of'_def projectKOs)
+  done
+
+lemma TCBSC_pred_map_state_refs_of':
+  "\<lbrakk>tcb_scs_of' s p = Some scp; pspace_aligned' s; pspace_distinct' s\<rbrakk>
+   \<Longrightarrow> (scp, TCBSchedContext) \<in> state_refs_of' s p"
+  apply (clarsimp simp: TCBSchedContext_state_refs_of' obj_at'_tcb_scs_of_equiv)
+  apply (clarsimp simp: tcb_of'_def)
+  apply (rule KOTCB_tcb_at'; simp)
+  apply (case_tac v'a; clarsimp)
+  done
+
+lemma SCTcb_pred_map_state_refs_of':
+  "\<lbrakk>scTCBs_of s scp = Some r; pspace_aligned' s; pspace_distinct' s\<rbrakk>
+   \<Longrightarrow> (r, SCTcb) \<in> state_refs_of' s scp"
+  apply (subgoal_tac "sc_at' scp s")
+   apply (clarsimp simp: state_refs_of'_def obj_at'_real_def ko_wp_at'_def projectKO_sc
+                         tcb_of'_Some)
+  apply (erule (1) KOSC_sc_at'[rotated])
+  apply (clarsimp simp: projectKO_sc)
+  done
+
+lemma SCReply_pred_map_state_refs_of':
+  "\<lbrakk>scReplies_of s p = Some r; pspace_aligned' s; pspace_distinct' s\<rbrakk>
+   \<Longrightarrow> (r, SCReply) \<in> state_refs_of' s p"
+  apply (subgoal_tac "sc_at' p s")
+   apply (clarsimp simp: state_refs_of'_def obj_at'_real_def ko_wp_at'_def projectKO_sc
+                         tcb_of'_Some)
+  apply (erule (1) KOSC_sc_at'[rotated])
+  apply (clarsimp simp: projectKO_sc)
+  done
+
+lemma ReplySC_pred_map_state_refs_of':
+  "\<lbrakk>replySCs_of s r = Some scp; pspace_aligned' s; pspace_distinct' s\<rbrakk>
+   \<Longrightarrow> (scp, ReplySchedContext) \<in> state_refs_of' s r"
+  apply (subgoal_tac "reply_at' r s")
+   apply (clarsimp simp: state_refs_of'_def obj_at'_real_def ko_wp_at'_def projectKO_reply)
+  apply (erule (1) KOReply_reply_at'[rotated])
+  apply (clarsimp simp: projectKO_reply)
+  done
+
+lemma invs'_tcbs_scs_sym_refs:
+  "\<lbrakk>sym_refs (state_refs_of' s); pspace_aligned' s; pspace_distinct' s; valid_objs' s\<rbrakk>
+   \<Longrightarrow> tcbs_scs_sym_refs s"
```

Comment:
I might be missing something, but we need `valid_objs’` here? Also, it would be nice to have some consistent names for this lemma and the next one.

Reply:
See also the comment below.